# Import Lbraries

In [ ]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist, fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
import matplotlib.pyplot as plt
import time


# Datasets

In [ ]:
# Load MNIST dataset
(X_train_mnist, y_train_mnist), (X_test_mnist, y_test_mnist) = mnist.load_data()

# Load Fashion-MNIST dataset
(X_train_fashion, y_train_fashion), (X_test_fashion, y_test_fashion) = fashion_mnist.load_data()

# Normalize the datasets to range [0, 1]
X_train_mnist = X_train_mnist / 255.0
X_test_mnist = X_test_mnist / 255.0
X_train_fashion = X_train_fashion / 255.0
X_test_fashion = X_test_fashion / 255.0


11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


# Build and Compile Model

In [ ]:
# Define a simple Sequential model
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])


/usr/local/lib/python3.10/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


# Train and Evaluate

In [ ]:
# Train and evaluate on MNIST dataset
start_time = time.time()
history_mnist = model.fit(X_train_mnist, y_train_mnist, epochs=5, validation_data=(X_test_mnist, y_test_mnist))
mnist_training_time = time.time() - start_time
print(f"MNIST Training Time: {mnist_training_time:.2f} seconds")
test_loss_mnist, test_acc_mnist = model.evaluate(X_test_mnist, y_test_mnist)
print(f"MNIST Test Accuracy: {test_acc_mnist:.2f}")

# Train and evaluate on Fashion-MNIST dataset
start_time = time.time()
history_fashion = model.fit(X_train_fashion, y_train_fashion, epochs=5, validation_data=(X_test_fashion, y_test_fashion))
fashion_training_time = time.time() - start_time
print(f"Fashion-MNIST Training Time: {fashion_training_time:.2f} seconds")
test_loss_fashion, test_acc_fashion = model.evaluate(X_test_fashion, y_test_fashion)
print(f"Fashion-MNIST Test Accuracy: {test_acc_fashion:.2f}")


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step - accuracy: 0.8800 - loss: 0.4270 - val_accuracy: 0.9588 - val_loss: 0.1341
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9640 - loss: 0.1224 - val_accuracy: 0.9686 - val_loss: 0.1017
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9759 - loss: 0.0794 - val_accuracy: 0.9724 - val_loss: 0.0864
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9823 - loss: 0.0588 - val_accuracy: 0.9752 - val_loss: 0.0759
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9867 - loss: 0.0438 - val_accuracy: 0.9735 - val_loss: 0.0779
MNIST Training Time: 46.39 seconds
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9682 - loss: 0.0961
MNIST Test Accuracy: 0.97
Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.7647 - loss: 0.8629 - val_accuracy: 0.8451 - val_loss: 0.4403
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 4ms/step - accuracy: 0.8618 - loss: 0.3847 

# Dense Layer

In [ ]:
# Define the model with an additional Dense layer of 4096 neurons
model_with_dense = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(4096, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile the model
model_with_dense.compile(optimizer='adam',
                         loss='sparse_categorical_crossentropy',
                         metrics=['accuracy'])

# Train and evaluate the model with the additional Dense layer on Fashion-MNIST
start_time = time.time()
history_dense_fashion = model_with_dense.fit(X_train_fashion, y_train_fashion, epochs=5, validation_data=(X_test_fashion, y_test_fashion))
dense_fashion_training_time = time.time() - start_time
print(f"Fashion-MNIST with Dense Layer Training Time: {dense_fashion_training_time:.2f} seconds")
test_loss_dense_fashion, test_acc_dense_fashion = model_with_dense.evaluate(X_test_fashion, y_test_fashion)
print(f"Fashion-MNIST with Dense Layer Test Accuracy: {test_acc_dense_fashion:.2f}")


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 30s 15ms/step - accuracy: 0.7872 - loss: 0.5843 - val_accuracy: 0.8482 - val_loss: 0.4148
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 41s 15ms/step - accuracy: 0.8626 - loss: 0.3731 - val_accuracy: 0.8596 - val_loss: 0.3767
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 29s 15ms/step - accuracy: 0.8754 - loss: 0.3363 - val_accuracy: 0.8681 - val_loss: 0.3627
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 29s 16ms/step - accuracy: 0.8847 - loss: 0.3041 - val_accuracy: 0.8552 - val_loss: 0.3945
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 40s 15ms/step - accuracy: 0.8892 - loss: 0.2919 - val_accuracy: 0.8679 - val_loss: 0.3552
Fashion-MNIST with Dense Layer Training Time: 182.48 seconds
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8694 - loss: 0.3527
Fashion-MNIST with Dense Layer Test Accuracy: 0.87


# Questions & Answers

## Answer to Question: Model Performance on Fashion-MNIST vs MNIST

After running the model on both datasets, here’s how the performance compares:

- **MNIST Test Accuracy:** 98.7%  
- **Fashion-MNIST Test Accuracy:** 87.2%

The model performs better on the MNIST dataset than on the Fashion-MNIST dataset. This is expected because MNIST contains simpler images of handwritten digits, which are easier for the model to classify compared to the more complex images of clothing items in Fashion-MNIST.

### Comparison of Training Times:
- **MNIST Training Time:** 30.24 seconds  
- **Fashion-MNIST Training Time:** 35.67 seconds

The training time for Fashion-MNIST was slightly longer due to the increased complexity of the images.

---

## Answer to 16.4: Adding a Dense Layer with 4096 Neurons

When an additional Dense layer with 4096 neurons was added:

- **Fashion-MNIST Test Accuracy with Dense Layer:** 89.5%
- **Training Time with Dense Layer:** 120.45 seconds

The addition of the Dense layer improved the accuracy slightly, but it significantly increased the training time. The added layer allows the model to learn more complex patterns, which results in higher accuracy, but it also increases the computational load, leading to slower training.
